src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -84,23 +65,151 @@ private[concurrent] object Promise {
     case t                                         => Failure(t)
   }
 
-   /**
-    * Latch used to implement waiting on a DefaultPromise's result.
-    *
-    * Inspired by: http://gee.cs.oswego.edu/cgi-bin/viewcvs.cgi/jsr166/src/main/java/util/concurrent/locks/AbstractQueuedSynchronizer.java
-    * Written by Doug Lea with assistance from members of JCP JSR-166
-    * Expert Group and released to the public domain, as explained at
-    * http://creativecommons.org/publicdomain/zero/1.0/
-    */
-    private final class CompletionLatch[T] extends AbstractQueuedSynchronizer with (Try[T] => Unit) {
-      override protected def tryAcquireShared(ignored: Int): Int = if (getState != 0) 1 else -1
-      override protected def tryReleaseShared(ignore: Int): Boolean = {
-        setState(1)
-        true
+  /* Encodes the concept of having callbacks.
+   * This is an `abstract method` to make sure calls are `invokevirtual` rather than `invokeinterface`
+   */
+  sealed abstract class Callbacks[+T] {
+    /* Logically prepends the callback `c` onto `this` callback */
+    def prepend[U >: T](c: Callbacks[U]): Callbacks[U]
+    /* Submits the callback function(s) represented by this Callback to be executed with the given value `v` */
+    def submitWithValue(v: Try[T @uncheckedVariance]): Unit
+  }
+
+  /* Represents 0 Callbacks, is used as an initial, sentinel, value for DefaultPromise
+   * This used to be a `case object` but in order to keep `Callbacks`'s methods bimorphic it was reencoded as `val`
+   */
+  val NoopCallback: Callbacks[Nothing] =
+    new Callback[Nothing](new ExecutionContext {
+      override def execute(r: Runnable): Unit = throw new IllegalStateException("Noop ExecutionContext.execute!")
+      override def reportFailure(t: Throwable): Unit = t.printStackTrace(System.err)
+      override def toString: String = "<Noop>"
+    }, _ => ())
+
+  /* Represents a single Callback function.
+     Precondition: `executor` is prepar()-ed */
+  final class Callback[T](
+    val executor: ExecutionContext,
+    val onComplete: Try[T] => Any) extends Callbacks[T] with Runnable with OnCompleteRunnable {
+
+    var value: Try[T] = null
```

Comment:
Just so I understand… this mutable variable is here so that we can treat this callback like a `Runnable`, but still have a way to pass a parameter to the callback. Is that right?